src/webgpu/api/operation/sampling/filter_mode.spec.ts

```diff
@@ -1,14 +1,629 @@
+    renderPass.draw(6);
+    renderPass.end();
+    t.device.queue.submit([commandEncoder.finish()]);
+
+    t.expectTexelViewComparisonIsOkInTexture(
+      { texture: renderTexture },
+      expectedColors(format, filterMode, addressModeU, addressModeV),
+      [kRenderSize, kRenderSize]
+    );
+  });
+
+g.test('minFilter')
+  .desc(
+    `
+  Test that for filterable formats, minFilter mode correctly modifies the sampling.
+    - format= {<filterable formats>}
+    - filterMode= {'nearest', 'linear'}
+    - addressModeU= {'clamp-to-edge', 'repeat', 'mirror-repeat'}
+    - addressModeV= {'clamp-to-edge', 'repeat', 'mirror-repeat'}
+  `
+  )
+  .params(u =>
+    u
+      .combine('format', kRenderableColorTextureFormats)
+      .filter(t => {
+        return (
+          kTextureFormatInfo[t.format].color.type === 'float' ||
+          kTextureFormatInfo[t.format].color.type === 'unfilterable-float'
+        );
+      })
+      .beginSubcases()
+      .combine('filterMode', kFilterModes)
+      .combine('addressModeU', kAddressModes)
+      .combine('addressModeV', kAddressModes)
+  )
+  .beforeAllSubcases(t => {
+    if (kTextureFormatInfo[t.params.format].color.type === 'unfilterable-float') {
+      t.selectDeviceOrSkipTestCase('float32-filterable');
+    }
+  })
+  .fn(t => {
+    const { format, filterMode, addressModeU, addressModeV } = t.params;
+    // For each pixel in the render result, a single sub-pixel sized quad is placed such that the
+    // center of the pixel corresponds to the region of the texture as if we are upsampling the
+    // texture.
+    const kRenderSize = 4;
+    const kUVs = filterMode === 'linear' ? kLinearUVs : kNearestUVs;
+
+    const sampler = t.device.createSampler({
+      addressModeU,
+      addressModeV,
+      minFilter: filterMode,
+    });
+    const sampleTexture = t.createTextureFromTexelView(
+      TexelView.fromTexelsAsColors(format, coord => {
+        const id = coord.x + coord.y * kCheckerTextureSize;
+        return kCheckerTextureData[id];
+      }),
+      {
+        size: [kCheckerTextureSize, kCheckerTextureSize],
+        usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST,
+      }
+    );
+    const renderTexture = t.device.createTexture({
+      format,
+      size: [kRenderSize, kRenderSize],
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+    });
+    const module = t.device.createShaderModule({
+      code: `
+      @group(0) @binding(0) var s : sampler;
+      @group(0) @binding(1) var t : texture_2d<f32>;
+
+      struct VertexOut {
+        @builtin(position) pos: vec4f,
+        @location(0) uv: vec2f,
+      };
+
+      @vertex
+      fn vs_main(@builtin(vertex_index) vi : u32,
+                 @builtin(instance_index) ii: u32) -> VertexOut {
+        const grid = vec2f(4., 4.);
+        const size = 1. / 2.;
+
+        const constantPlaneOffset = vec2f(3. / 8., 3. / 8.);
+        const perPixelOffset = -1. / 4.;
```

Comment:
it's not obvious _why_ these values are the way they are without seeing a picture what what we're trying to do - importantly how you're placing the smaller sub-pixel quad inside the pixel.

Maybe you can have a small ascii art picture to depict the goal

Reply:
Done